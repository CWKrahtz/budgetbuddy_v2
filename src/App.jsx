import React, { useState } from 'react'
import Col from 'react-bootstrap/Col'
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import './App.css'
import Income from './components/Income'
import Expenses from './components/Expenses'
import Taxes from './components/Taxes'
import TotalCard from './components/items/TotalCard'
import { dummyCards, lastTotalCard } from './utils'
import Savings from './components/Savings'
import LastTotalCard from './components/items/LastTotalCard'
import { Button } from 'react-bootstrap'
import { addIncome, calculateSave, calculateTax } from './functions/income'
// import {addIncome , calculateTax} from './functions/income'  

function App() {

  const [incomes, setIncomes] = useState([])

  const [expenses, setExpenses] = useState([])

  const handleAddingNewIncome = (newIncome) => {
    // console.log('Begin of handle')
    var newIncomeList = calculateTax(newIncome)
    // newIncomeList = addIncome(incomes, newIncomeList)
    setIncomes([...incomes, newIncomeList])
    // console.log('End of handle')
  }
  
  const handleAddingNewExppence = (newExpense) => {
    setExpenses([...expenses, newExpense])
  }
  

  return (
    <Container fluid>
      <Row >
        <Col xs={12} className='topbar p-3 mb-4'>
          <h2 className='title'>
            <img src='logo512.png' width={60} height={60} />
            <span>BudgetBuddy</span>
          </h2>
          <Button><img src='reset.png' width={30} height={30} /></Button>
        </Col>
      </Row>
      <Row>
        <Col xs={12} sm={12} md={5}>
          <div className='card shadow-sm p-3 mb-4 bg-white rounded'>
            <Income incomes={incomes} handleAddNew={handleAddingNewIncome} />
          </div>
          <div className='card shadow-sm p-3 mb-4 bg-white rounded'>
            <Taxes incomes={incomes}/>
          </div>
          <div className='card shadow-sm p-3 mb-4 bg-white rounded'>
            <Savings incomes={incomes}/>
          </div>
        </Col>
        <Col xs={12} md={4} >
          <div className='card shadow-sm p-3 mb-4 bg-white rounded'>
            <Expenses expenses={expenses} handleAddingNewE={handleAddingNewExppence}/>
          </div>
        </Col>

        <Col xs={12} md={3}>
          {dummyCards.map((total, index) => (
            <div key={index} className='card shadow-sm p-3 mb-4 rounded'>
              <TotalCard total={total} incomes={incomes} expenses={expenses}/>
            </div>
          ))}
          <div className='card shadow-sm p-3 mb-4 rounded last-total-card'>
            <LastTotalCard total={lastTotalCard} />
          </div>
        </Col>
      </Row>
    </Container>
  )
}

export default App
