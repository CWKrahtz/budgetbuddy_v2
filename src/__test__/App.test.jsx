import { render, screen } from "@testing-library/react"
import App from "../App"
import userEvent from "@testing-library/user-event"


test("Snapshot of app when redered first time", () => {

    render(<App />)

    var app = screen.getByTestId("outer-app")

    expect(app).toBeInTheDocument()

    expect(app).toMatchSnapshot()

})

test("snapshot of when new item has been added", async () => {

    var user = userEvent.setup()

    //ARRANGE
    render(<App />)

    var app = screen.getByTestId("outer-app")
    
    var name = screen.getByLabelText("income-name")
    await user.click(name)
    await user.keyboard("Test Text")
    expect(name).toHaveValue("Test Text")

    var income = screen.getByLabelText("income")
    await user.click(income)
    await user.keyboard("5")
    expect(income).toHaveValue(5)

    var icon = screen.getByLabelText("income-icon")
    await user.click(icon)
    await user.selectOptions(icon, "üê∂")
    expect(icon).toHaveValue("üê∂")

    var button = screen.getByText("Add Income") 
    await user.click(button)

    // //ASSERT
    expect(app).toMatchSnapshot()

})

//1, if Not snapshot == create a new one
//2. if exist, create new one, and compare it to the previous one
//   if fail - something changed
// To update snapshot: npm run test -- -u           (not working)
// To update Snapshot: npx jest --updateSnapshot    (working)