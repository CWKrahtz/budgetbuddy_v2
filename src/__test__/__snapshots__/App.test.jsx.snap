// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of app when redered first time 1`] = `
<div
  class="container-fluid"
  data-testid="outer-app"
>
  <div
    class="row"
  >
    <div
      class="topbar p-3 mb-4 col-12"
    >
      <h2
        class="title"
      >
        <img
          height="60"
          src="logo512.png"
          width="60"
        />
        <span>
          BudgetBuddy
        </span>
      </h2>
      <button
        class="btn btn-primary"
        type="button"
      >
        <img
          height="30"
          src="reset.png"
          width="30"
        />
      </button>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col-md-5 col-sm-12 col-12"
    >
      <div
        class="card shadow-sm p-3 mb-4 bg-white rounded"
      >
        <div>
          <h3>
            Household Montly Salaries
          </h3>
          <div
            class="form-row"
          >
            <select
              aria-label="income-icon"
              autocomplete="off"
              class="form-select"
              data-testid="icon"
              id="icon"
              name="icon"
            >
              <option>
                -
              </option>
              <option
                value="üë©"
              >
                üë©
              </option>
              <option
                value="üë±‚Äç‚ôÇÔ∏è"
              >
                üë±‚Äç‚ôÇÔ∏è
              </option>
              <option
                value="üê∂"
              >
                üê∂
              </option>
              <option
                value="üê±"
              >
                üê±
              </option>
              <option
                value="üê∞"
              >
                üê∞
              </option>
              <option
                value="üêª"
              >
                üêª
              </option>
              <option
                value="ü¶Å"
              >
                ü¶Å
              </option>
              <option
                value="üê∑"
              >
                üê∑
              </option>
              <option
                value="üêÆ"
              >
                üêÆ
              </option>
            </select>
            <input
              aria-label="income-name"
              autocomplete="off"
              class="form-control"
              data-testid="name"
              id="name"
              name="name"
              placeholder="Member Name"
              type="text"
              value=""
            />
            <input
              aria-label="income"
              autocomplete="off"
              class="form-control"
              data-testid="income"
              id="income"
              name="income"
              placeholder="0.00"
              step="0.01"
              type="number"
              value=""
            />
            <button
              aria-label="income-button"
              class="add-income btn btn-primary"
              type="button"
            >
              Add Income
            </button>
          </div>
          <div
            class="income-outer hide-scroll"
          />
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 bg-white rounded"
      >
        <div
          data-toggle="tooltip"
          title="Those bastards!"
        >
          <h3>
            Tax Bracket Calculation
          </h3>
          <div
            class="scroll-row hide-scroll"
          />
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 bg-white rounded"
      >
        <div
          data-testid="save-card"
        >
          <div
            class="title-row"
          >
            <h3>
              Savings Calculation
            </h3>
            <span>
              <p>
                % you want to save
              </p>
              <select
                aria-label="percentage"
                autocomplete="off"
                class="form-select"
                id="percentage"
                name="percentage"
              >
                <option
                  selected=""
                >
                  -
                </option>
                <option
                  value="5"
                >
                  5
                  %
                </option>
                <option
                  value="10"
                >
                  10
                  %
                </option>
                <option
                  value="15"
                >
                  15
                  %
                </option>
                <option
                  value="20"
                >
                  20
                  %
                </option>
              </select>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-md-4 col-12"
    >
      <div
        class="card shadow-sm p-3 mb-4 bg-white rounded"
      >
        <div>
          <h3>
            Household Montly Expenses
          </h3>
          <div
            class="form-row"
          >
            <input
              aria-label="name"
              autocomplete="off"
              class="form-control"
              id="title"
              name="title"
              placeholder="Expense Title"
              type="text"
              value=""
            />
            <input
              aria-label="expense"
              autocomplete="off"
              class="form-control"
              id="expense"
              name="expense"
              placeholder="0.00"
              step="0.01"
              type="number"
              value=""
            />
            <button
              class="add-expenses btn btn-primary"
              type="button"
            >
              Add
            </button>
          </div>
          <div
            class="expense-outer hide-scroll"
          />
        </div>
      </div>
    </div>
    <div
      class="col-md-3 col-12"
    >
      <div
        class="card shadow-sm p-3 mb-4 rounded"
      >
        <div
          class="total-card p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(228, 253, 205);"
          >
            <img
              height="40"
              src="income.png"
              width="40"
            />
          </div>
          <div
            class="labels"
          >
            <h4>
              R 
              0.00
            </h4>
            <p>
              Total Income Before Tax
            </p>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 rounded"
      >
        <div
          class="total-card p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(255, 238, 210);"
          >
            <img
              height="40"
              src="tax.png"
              width="40"
            />
          </div>
          <div
            class="labels"
          >
            <h4>
              R 
              0.00
            </h4>
            <p>
              Total Income After Tax
            </p>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 rounded"
      >
        <div
          class="total-card p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(255, 226, 211);"
          >
            <img
              height="40"
              src="expense.png"
              width="40"
            />
          </div>
          <div
            class="labels"
          >
            <h4>
              R 
              0.00
            </h4>
            <p>
              Total Expenses
            </p>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 rounded"
      >
        <div
          class="total-card p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(221, 221, 247);"
          >
            <img
              height="40"
              src="savings.png"
              width="40"
            />
          </div>
          <div
            class="labels"
          >
            <h4>
              R 
              0.00
            </h4>
            <p>
              Total Savings
            </p>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 rounded last-total-card"
      >
        <div
          class="final-total p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(207, 242, 109);"
          >
            <img
              height="40"
              src="wallet.png"
              width="40"
            />
          </div>
          <h2>
            R 
            1.00
          </h2>
          <p>
            Total Left To Spend Each Month üò≠
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshot of when new item has been added 1`] = `
<div
  class="container-fluid"
  data-testid="outer-app"
>
  <div
    class="row"
  >
    <div
      class="topbar p-3 mb-4 col-12"
    >
      <h2
        class="title"
      >
        <img
          height="60"
          src="logo512.png"
          width="60"
        />
        <span>
          BudgetBuddy
        </span>
      </h2>
      <button
        class="btn btn-primary"
        type="button"
      >
        <img
          height="30"
          src="reset.png"
          width="30"
        />
      </button>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col-md-5 col-sm-12 col-12"
    >
      <div
        class="card shadow-sm p-3 mb-4 bg-white rounded"
      >
        <div>
          <h3>
            Household Montly Salaries
          </h3>
          <div
            class="form-row"
          >
            <select
              aria-label="income-icon"
              autocomplete="off"
              class="form-select"
              data-testid="icon"
              id="icon"
              name="icon"
            >
              <option>
                -
              </option>
              <option
                value="üë©"
              >
                üë©
              </option>
              <option
                value="üë±‚Äç‚ôÇÔ∏è"
              >
                üë±‚Äç‚ôÇÔ∏è
              </option>
              <option
                value="üê∂"
              >
                üê∂
              </option>
              <option
                value="üê±"
              >
                üê±
              </option>
              <option
                value="üê∞"
              >
                üê∞
              </option>
              <option
                value="üêª"
              >
                üêª
              </option>
              <option
                value="ü¶Å"
              >
                ü¶Å
              </option>
              <option
                value="üê∑"
              >
                üê∑
              </option>
              <option
                value="üêÆ"
              >
                üêÆ
              </option>
            </select>
            <input
              aria-label="income-name"
              autocomplete="off"
              class="form-control"
              data-testid="name"
              id="name"
              name="name"
              placeholder="Member Name"
              type="text"
              value=""
            />
            <input
              aria-label="income"
              autocomplete="off"
              class="form-control"
              data-testid="income"
              id="income"
              name="income"
              placeholder="0.00"
              step="0.01"
              type="number"
              value=""
            />
            <button
              aria-label="income-button"
              class="add-income btn btn-primary"
              type="button"
            >
              Add Income
            </button>
          </div>
          <div
            class="income-outer hide-scroll"
          >
            <div
              data-testid="income-card"
            >
              <div
                class="card card-row p-2 mb-2 row-income"
              >
                <span>
                  <div
                    class="avatar"
                  >
                    üê∂
                  </div>
                  <p>
                    1
                    . 
                    Test Text
                  </p>
                </span>
                <p
                  class="amount"
                >
                  R 
                  5.00
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 bg-white rounded"
      >
        <div
          data-toggle="tooltip"
          title="Those bastards!"
        >
          <h3>
            Tax Bracket Calculation
          </h3>
          <div
            class="scroll-row hide-scroll"
          >
            <div
              data-testid="taxes-card"
            >
              <div
                class="card card-col p-2 row-tax"
              >
                <div
                  class="card-row"
                >
                  <span>
                    <div
                      class="avatar"
                    >
                      üê∂
                    </div>
                    <p>
                      Test Text
                    </p>
                  </span>
                  <h4>
                    18
                    %
                  </h4>
                </div>
                <p
                  class="amount"
                >
                  - R 
                  0.00
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 bg-white rounded"
      >
        <div
          data-testid="save-card"
        >
          <div
            class="title-row"
          >
            <h3>
              Savings Calculation
            </h3>
            <span>
              <p>
                % you want to save
              </p>
              <select
                aria-label="percentage"
                autocomplete="off"
                class="form-select"
                id="percentage"
                name="percentage"
              >
                <option
                  selected=""
                >
                  -
                </option>
                <option
                  value="5"
                >
                  5
                  %
                </option>
                <option
                  value="10"
                >
                  10
                  %
                </option>
                <option
                  value="15"
                >
                  15
                  %
                </option>
                <option
                  value="20"
                >
                  20
                  %
                </option>
              </select>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-md-4 col-12"
    >
      <div
        class="card shadow-sm p-3 mb-4 bg-white rounded"
      >
        <div>
          <h3>
            Household Montly Expenses
          </h3>
          <div
            class="form-row"
          >
            <input
              aria-label="name"
              autocomplete="off"
              class="form-control"
              id="title"
              name="title"
              placeholder="Expense Title"
              type="text"
              value=""
            />
            <input
              aria-label="expense"
              autocomplete="off"
              class="form-control"
              id="expense"
              name="expense"
              placeholder="0.00"
              step="0.01"
              type="number"
              value=""
            />
            <button
              class="add-expenses btn btn-primary"
              type="button"
            >
              Add
            </button>
          </div>
          <div
            class="expense-outer hide-scroll"
          />
        </div>
      </div>
    </div>
    <div
      class="col-md-3 col-12"
    >
      <div
        class="card shadow-sm p-3 mb-4 rounded"
      >
        <div
          class="total-card p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(228, 253, 205);"
          >
            <img
              height="40"
              src="income.png"
              width="40"
            />
          </div>
          <div
            class="labels"
          >
            <h4>
              R 
              5.00
            </h4>
            <p>
              Total Income Before Tax
            </p>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 rounded"
      >
        <div
          class="total-card p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(255, 238, 210);"
          >
            <img
              height="40"
              src="tax.png"
              width="40"
            />
          </div>
          <div
            class="labels"
          >
            <h4>
              R 
              5.00
            </h4>
            <p>
              Total Income After Tax
            </p>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 rounded"
      >
        <div
          class="total-card p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(255, 226, 211);"
          >
            <img
              height="40"
              src="expense.png"
              width="40"
            />
          </div>
          <div
            class="labels"
          >
            <h4>
              R 
              0.00
            </h4>
            <p>
              Total Expenses
            </p>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 rounded"
      >
        <div
          class="total-card p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(221, 221, 247);"
          >
            <img
              height="40"
              src="savings.png"
              width="40"
            />
          </div>
          <div
            class="labels"
          >
            <h4>
              R 
              0.00
            </h4>
            <p>
              Total Savings
            </p>
          </div>
        </div>
      </div>
      <div
        class="card shadow-sm p-3 mb-4 rounded last-total-card"
      >
        <div
          class="final-total p-2"
        >
          <div
            class="icon"
            style="background-color: rgb(207, 242, 109);"
          >
            <img
              height="40"
              src="wallet.png"
              width="40"
            />
          </div>
          <h2>
            R 
            1.00
          </h2>
          <p>
            Total Left To Spend Each Month üò≠
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
